src/process.rs

```diff
@@ -1,8 +1,224 @@
+    // API in unit tests
+    reviews: RemoteData<Vec<ReviewState>>,
+    statuses: RemoteData<HashMap<String, StatusState>>,
+}
+
+impl<'a> Analyzer<'a> {
+    pub fn new(pr: &'a PR, client: &'a context::Client, config: &'a context::Config) -> Self {
+        Self {
+            pr,
+            client,
+            config,
+            reviews: RemoteData::Remote,
+            statuses: RemoteData::Remote,
+        }
+    }
+
+    /// Analyze a PR to determine what actions need to be undertaken.
+    pub async fn required_actions(&self) -> Result<Actions> {
+        let pr = &self.pr;
+        let mut actions = Actions::noop();
+
+        if pr.draft {
+            log::info!("PR #{}: Draft, nothing to do", pr.number);
+            return Ok(actions);
+        }
+
+        if pr.state == IssueState::Closed {
+            log::info!("PR #{}: Closed, nothing to do", pr.number);
+            return Ok(actions);
+        }
+
+        if pr.updated_at < Utc::now() - Duration::minutes(60) {
+            log::info!("PR #{}: Inactive for over 1 hour, nothing to do", pr.number);
+            return Ok(actions);
+        }
+
+        // Now that the basic checks have been passed we can gather information
+        // from the GitHub API in order to do the full check. We do this second
+        // so that we use the GitHub API as little as possible, we don't want to
+        // hit the rate limit.
+        let statuses_passed = self.pr_statuses_passed().await?;
+        let pr_approved = self.pr_approved().await?;
+        let mut description_ok = true;
+
+        // Assign the "reviewed" label if there is one and the PR is approved
+        if let Some(label) = self.config.reviewed_label.clone() {
+            actions.set_label(label, pr_approved);
+        }
+
+        // Assign the "needs-description" label if there is one and the PR lacks one
+        if let Some(label) = self.config.needs_description_label.clone() {
+            description_ok = self.pr.has_description;
+            actions.set_label(label, !self.pr.has_description);
+        }
+
+        // Assign the "ci-passed" label if CI passed
+        actions.set_label(self.config.ci_passed_label.to_string(), statuses_passed);
+
+        actions.set_merge(
+            !self.block_merge_label_applied() && description_ok && pr_approved && statuses_passed,
+        );
+
+        return Ok(actions);
+    }
+
+    async fn pr_approved(&self) -> Result<bool> {
+        let review_not_required = self.config.reviewed_label.is_none();
+        let mut waiting = false;
+        let mut approved = review_not_required;
+        for review in self.get_pr_reviews().await?.iter() {
+            match review {
+                ReviewState::Approved => approved = review_not_required || true,
+                ReviewState::Pending | ReviewState::ChangesRequested => waiting = true,
+                _ => (),
+            }
+        }
+        Ok(approved && !waiting)
```

Comment:
I don't think this logic is correct, when you get the review state of a PR it's essentially a log of all review activity, which means if a reviewer requested changes, then the author resolved them, and the reviewer then approved the change, this would not mark the PR as reviewed, even though the state of the last review from that user was approved.

Reply:
Oh! I think I have got confused about how reviews are represented in the API. Thanks Jake!